### ISSUE TITLE
Find a way to store / read map-like structures

### ISSUE THREAD

**Issue by OlivierBondu** [cp3-llbb/Framework#81]
To track the various discussions last week with @blinkseb @swertz and @BrieucF :
- so far we store all of our favorite end-candidate (hh, tt, za, etc) with their associated quality (lepton id tight, lepton iso loose, btag medium, highest ptjj, etc)
- it would be nice to have a way to access them by their quality (easier / safer use of anything running on the produced out-trees)

the solutions envisaged were:
- store a map "myfavoritecombination" to "vector of candidates": easy to code, but unfortunately does not work out of the box e.g. with the plotter (`Warning in <TSelectorDraw::ProcessFillObject>: Not implemented for map<int,vector<unsigned int> >`): it seems the only allowed keys in bare root are TBits (even now that we do have access to the struct dictionnaries).
  - the workaround would be to modify root
- store a function taking as inputs the different object qualities 
  - this would need to work a bit on the plotter and other scripts to call it, see for example [this discussion](https://root.cern.ch/phpBB3/viewtopic.php?t=11193)
- store a vector of 'vector of candidates'
  - error prone, need to open the corresponding header / code to make sure what is called when accessing `v[i]`, probably not sustainable in the long run
- store a struct for each possible combination
  - very inefficient disk-wise, not sustainable in the long run

comments are very welcome :)

**Comment by blinkseb** [comment]
A few comments:
- maps are not handled at in by ROOT in the `Draw` function. Basically, `TTreeFormula` parses the string, and tries to identify the return type of each operand, which must be a number or a string. In the case of the map, `TTreeFormula` assumes that the return type is an object, and it fails, because the only object supported by `TTreeFormula` is a `TBits`. As Olivier stated, it should be fixable fairly easily, as the logic should already be in place to handle `std::vector`. But it means changing ROOT, and so will only work locally (grid: out)
- It's not possible to store function inside a ROOT file. What you could do is executing an user script which can in turn call arbitrary code. Maybe we can think of a way to combine #2 and #3: everything is stored as a vector of vector inside the tree, but we do not access it directly, but via a function call which return the appropriate vector. Not sure how easy it's to implement.

**Comment by swertz** [comment]
I've tried the "vector of vectors" solution, and I think it gets much less error prone when using enumerations to remember indices, for instance:

```
enum class bTagWP{ L, M, T, Count };
std::vector<std::vector<int>> selectedBJets(size_t(bTagWP::Count), std::vector<int>);
myJets[index].bTag[size_t(bTagWP::L)] = true;
selectedBJets[size_t(bTagWP::L)].push_back(index);
```

And similarly for e.g. LL, LM, ... IDs. It also allows to iterate over all working points to automatize some stuff (which comes in handy for dileptons/dijets, where there are 9 different pairs of IDs). It's just a bit ugly because of the size_t() cast necessary to call std::vector::operator[](maybe there's a way to bypass that?).
Then, the C++ header file where the enum class is defined can be imported in the python script used to generate the histograms (gROOT.ProcessLine("#include blablah."))...
Of course, for size reasons only one vector of Lepton/DiLepton/... structures should be stored, while vectors of vectors can then be used to store specific collections of indices to the aforementioned vector of structures.

Another way would be using #defines, which can then still be imported in python. And it is still possible to create an array on which to iterate...

**Comment by blinkseb** [comment]
I think what @swertz proposes is the easiest solution and yet powerful enough for our needs. We can even avoid the cast by defining the enum as

``` C++
namespace bTagWP {
    enum bTagWP { L, M, T, Count };
};
```

You gain implicit `int` conversion for the enum.

**Comment by swertz** [comment]
To use the enums in the python file I would then do somethig like that, unless someone has a better idea:
In the header.h file:

```
namespace bTagWP {
    enum bTagWP { L, M, T };
}
```

In the python file:

```
import ROOT as R
R.gROOT.ProcessLine("#include \"header.h\"")
myBTagWP = { "L": bTagWP.L, "M": bTagWP.M, "T": bTagWP.T }
```

Which allows to loop over myBTagWP, accessing the correct indices in the Tree and modifying the histogram names & titles according to the working points.

**Comment by swertz** [comment]
Interesting suggestion by @OlivierBondu : using a single index to combine different informations (ID, iso, ...) for one or several objects, ie.:

```
  L M T
L 0 1 2
T 3 4 5
```

This will require
- a function returning the correct index for each combination of information/objects
- a way to iterate over all or a specific subset of indices

**Comment by OlivierBondu** [comment]
if we define something like:

``` C++
int bitA = lepIso::Count;
int bitB = bitA * lepID::Count;
int bitC = bitB * lepIso::Count;
```

then we should be able to access the correct vector of indices (example with lep1 ID Loose, lep1 ID tight, lep2 ID Loose, lep2 Isi Loose):

``` C++
ll_id_iso[lepID::L * bitC + lepIso::T * bitB + lepID::L * bitA + lepIso::L]
```

the painful thing to write is filling up the map itself, but usage should be quite straightforward

**Comment by swertz** [comment]
Indeed, so the functions would look like

```
int LepIDIsoLepIDIso(LepID id1, LepIso iso1, LepID id2, LepIso is2) { }
```

And iterations would work just as before, only that the function would need to be called inside the loop:

```
for(const LepID& id1: LepID::iterator){
 for(const LepID& id2: LepID::iterator){
  int idx = LepIDIsoLepIDIso(id1, LepIso::T, id2, LepIso::L);
  const DiLepton& m_diLepton = diLeptons[ diLeptons_ID_Iso[idx] ];
 }
}
```

Question: how do you suggest to handle the asymmetry between electrons (no specific Iso, but also Veto ID) and muons, since we would like to handle leptons altogether?

**Comment by OlivierBondu** [comment]
all seems to work in [HH PR 29](https://github.com/cp3-llbb/HHAnalysis/pull/29)

one thing that is still missing though what we already discussed at [HH issue 18](https://github.com/cp3-llbb/HHAnalysis/issues/18): we still would need to store the nEntries of each vector (no easy way around I'm afraid)

**Comment by swertz** [comment]
Why is that? Draw("leptons_ID_Iso[the index I want].size()","","") would still work even with the fancy multi-index right?

**Comment by swertz** [comment]
So, it seems even 3D vectors are not supported by TTree::Draw() (unless I'm doing something wrong): accessing "tt_diBJets_DRCut_BBWPs_PtOrdered[0][0][0]" in /home/fynu/swertz/scratch/CMSSW_7_4_10/src/cp3_llbb/TTAnalysis/test/output_mc.root doesn't work.

Using TTree::Draw() is like navigating through a maze of things you would expect to work but don't...

**Comment by blinkseb** [comment]
:sad:

**Comment by swertz** [comment]
Another issue: for an object combining two leptons and two jets, assuming we want every combination of lepton ID, lepton isolation, jet ID and jet b-tagging, that gives about 2900 entries... A bit much?
If we fix the jet ID and the muon isolation for now, that reduces to a more manageable 81 entries.

**Comment by OlivierBondu** [comment]
@swertz : see https://github.com/cp3-llbb/HHAnalysis/issues/18, getting the size of the vector inside a vector does not seem to work

As for the number of entries, we have to look at the size... it may be compressed pretty well in the end ?

**Comment by swertz** [comment]
It depends what we're talking about... 
I've done a quick test on a `vector<vector>`:
- `@v.size()` returns the size of the first component of the vector
- `@v[0].size()` seems to return the same thing (ie not what we want)
- `v[0].size()` returns the size of the second component of the first entry of the vector

So if you want to know the number of loose leptons, and 0 corresponds to loose, `lepton_idxs[0].size()` should work right (if `lepton_idxs` is the 2D vector pointing to entries in the `vector<Lepon>` collection)?

**Comment by OlivierBondu** [comment]
well I can't seem to have it work:

``` C++
root [15] t->Draw("hh_map_l_id_iso[2].size()")
Warning in <TTreeFormula::DefinedVariable>: Can not call method on content of vector<int> in hh_map_l_id_iso[2].size()

Error in <TTreeFormula::Compile>:  Part of the Variable "hh_map_l_id_iso[2].size()" exists but some of it is not accessible or useable
Info in <TSelectorDraw::AbortProcess>: Variable compilation failed: {hh_map_l_id_iso[2].size(),}
```

how did you test this ?

**Comment by swertz** [comment]
No you're right it doesn't work... What I had tried is for a 3D vector:

```
t->Draw("tt_diLepDiBJets_DRCut_BBWPs_PtOrdered[0][0].size()","","")
```

Which for whatever reason didn't crash, but I'm not sure if the output is correct either. Anyway, since we can't access the third component, this is moot.

What I've also tried is `Draw("Length$(vector[0][])","","")` which _should_ work, since this is supposed to return the number of iterations over free indices in the formula - here exactly what we want! But it doesn't... I've asked a question on RootTalk, let's see what it gives.
